plugins/suspendmanager.cpp

```diff
@@ -0,0 +1,822 @@
+                                is_enabled ? "enabled" : "disabled",
+                                is_enabled ? "enabled" : "disabled");
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown (color_ostream &out) {
+    DEBUG(control,out).print("shutting down %s\n", plugin_name);
+    suspendmanager_instance.release();
+    eventhandler_instance.release();
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_load_site_data (color_ostream &out) {
+    cycle_timestamp = 0;
+    config = World::GetPersistentSiteData(CONFIG_KEY);
+
+    if (!config.isValid()) {
+        DEBUG(control,out).print("no config found in this save; initializing\n");
+        config = World::AddPersistentSiteData(CONFIG_KEY);
+        config.set_bool(CONFIG_IS_ENABLED, is_enabled);
+    }
+
+    is_enabled = config.get_bool(CONFIG_IS_ENABLED);
+    DEBUG(control,out).print("loading persisted enabled state: %s\n",
+                            is_enabled ? "true" : "false");
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (event == DFHack::SC_WORLD_UNLOADED) {
+        if (is_enabled) {
+            DEBUG(control,out).print("world unloaded; disabling %s\n",
+                                    plugin_name);
+            is_enabled = false;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    if (is_enabled && world->frame_counter - cycle_timestamp >= CYCLE_TICKS)
+        do_cycle(out);
+    return CR_OK;
+}
+
+static command_result do_command(color_ostream &out, vector<string> &parameters) {
+    // be sure to suspend the core if any DF state is read or modified
+    CoreSuspender suspend;
+
+    if (!Core::getInstance().isMapLoaded() || !World::IsSiteLoaded()) {
+        out.printerr("Cannot run %s without a loaded fort.\n", plugin_name);
+        return CR_FAILURE;
+    }
+
+    if (parameters[0] == "now") {
+        do_cycle(out);
+        return CR_OK;
+    } else if (parameters[0] == "enable") {
+        return plugin_enable(out,true);
+    } else if (parameters[0] == "disable") {
+        return plugin_enable(out,false);
+    } else if (parameters[0] == "set" && parameters[1] == "preventblocking") {
+        if (parameters[3] == "true") {
+            suspendmanager_instance->prevent_blocking = true;
+            if (is_enabled) do_cycle(out);
+            return CR_OK;
+        } else if (parameters[3] == "false") {
+            suspendmanager_instance->prevent_blocking = false;
+            if (is_enabled) do_cycle(out);
+            return CR_OK;
+        } else
+            return CR_WRONG_USAGE;
+    } else {
+        return CR_WRONG_USAGE;
+    }
+
+}
+
+static void jobCompletedHandler(color_ostream& out, void* ptr) {
+    DEBUG(cycle,out).print("job completed; updating suspensions\n");
+    df::job* job = static_cast<df::job*>(ptr);
+    if (SuspendManager::isConstructionJob(job))
```

Comment:
or a smooth/carving job

Reply:
Completed smoothing/carving jobs, unlike new walls, do not risk blocking other jobs. So I think it's fine to not refresh immediately and wait for the next job- or time-triggered cycle.